Lib/test/test_unparse.py

```diff
@@ -130,7 +130,43 @@ class Foo: pass
 
 class ASTTestCase(unittest.TestCase):
     def assertASTEqual(self, ast1, ast2):
-        self.assertEqual(ast.dump(ast1), ast.dump(ast2))
+        # Ensure the comparisons start at an AST node
+        self.assertIsInstance(ast1, ast.AST)
+        self.assertIsInstance(ast2, ast.AST)
+
+        # An AST comparison routine modeled after ast.dump(), but
+        # instead of string building, it traverses the two trees
+        # in lock-step.
+        def traverse_compare(a, b, missing=object()):
```

Comment:
Since you moved missing inside the function definition, now you can move the function definition at the class level (method) of the module level (function, you should pass the testcase as an argument in this case), to avoid the cost of defining a new function at each assertASTEqual() call.

Reply:
If you are concerned about performance, using an inner function is **faster** than calling a method.  True, the setup cost is greater, but that time is more than recovered due to attribute lookup after a few hundred comparisons.  Global (module level) functions suffer the same fate as well but require more comparisons before getting overtaken.

After instrumenting test_unparse to get the number of calls to assertASTEqual and compares, I whipped up a reproducible minimal test script to show the timings.  The script is [here](https://gist.github.com/jkloth/d5ab8540e5e47d29990321548e79c59e) for you to run yourself.

Using main on Win10:
100 -> inner: 29 usec method: 27.4 usec
200 -> inner: 56.8 usec method: 48 usec
300 -> inner: 76.3 usec method: 86.4 usec
CosmeticTestCase (14 compares)
  inner: 88.9 usec
  method: 77.9 usec
DirectoryTestCase (400 compares)
  inner: 74.1 msec
  method: 79.5 msec
UnparseTestCase (400 compares)
  inner: 91 msec
  method: 97.6 msec

Reply:
As I wrote previously, if there is a good reason to define a "local" (nested?) function, please add a comment here for future readers. Otherwise, someone who doesn't like nested functions can move the function (during a random refactoring) without knowing that you ran benchmarks and the code is faster when written like that. It's very easy to lose information when code is read/modified 5, 10 or 20 years later. And yes, it happens to me often to read 20 years old Python code and I have to dig into Git history (with unpleasant issues about CVS, SVN and HG commit numbers) and very old bug tracker issues to attempt to rebuild the rationale for the existing code before feeling safe to change it.

The performance doesn't make sense (method/module level code) to me. It should be as fast or faster. But my only god are benchmarks: I only trust benchmarks numbers :-)